rclc_parameter/src/rclc_parameter/parameter_server.c

```diff
@@ -827,6 +1545,134 @@ rclc_parameter_server_init_service(
+    return RCL_RET_ERROR;
+  }
+
+  // Set constrain description
+  if (!rclc_parameter_set_string(
+      &parameter_descriptor->additional_constraints,
+      additional_constraints))
+  {
+    return RCL_RET_ERROR;
+  }
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rclc_set_parameter_read_only(
+  rclc_parameter_server_t * parameter_server,
+  const char * parameter_name, bool read_only)
+{
+  if (parameter_server->on_callback) {
+    return RCLC_PARAMETER_DISABLED_ON_CALLBACK;
+  }
+
+  size_t index = rclc_parameter_search_index(&parameter_server->parameter_list, parameter_name);
+
+  if (index >= parameter_server->parameter_list.size) {
+    return RCL_RET_ERROR;
+  }
+
+  ParameterDescriptor * parameter_descriptor = &parameter_server->parameter_descriptors.data[index];
+
+  // Set flag
+  parameter_descriptor->read_only = read_only;
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rclc_add_parameter_constraint_double(
+  rclc_parameter_server_t * parameter_server,
+  const char * parameter_name, double from_value,
+  double to_value, double step)
+{
+  if (parameter_server->on_callback) {
+    return RCLC_PARAMETER_DISABLED_ON_CALLBACK;
+  }
+
+  size_t index = rclc_parameter_search_index(&parameter_server->parameter_list, parameter_name);
+
+  if (index >= parameter_server->parameter_list.size) {
+    return RCL_RET_ERROR;
+  }
+
+  ParameterDescriptor * parameter_descriptor = &parameter_server->parameter_descriptors.data[index];
+
+  if (parameter_descriptor->type != RCLC_PARAMETER_DOUBLE) {
+    return RCL_RET_ERROR;
+  }
+
+  parameter_descriptor->floating_point_range.data->from_value = from_value;
+  parameter_descriptor->floating_point_range.data->to_value = to_value;
+  parameter_descriptor->floating_point_range.data->step = step;
+  parameter_descriptor->floating_point_range.size = 1;
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rclc_add_parameter_constraint_integer(
+  rclc_parameter_server_t * parameter_server,
+  const char * parameter_name, int64_t from_value,
+  int64_t to_value, uint64_t step)
+{
+  if (parameter_server->on_callback) {
+    return RCLC_PARAMETER_DISABLED_ON_CALLBACK;
+  }
+
+  size_t index = rclc_parameter_search_index(&parameter_server->parameter_list, parameter_name);
+
+  if (index >= parameter_server->parameter_list.size) {
+    return RCL_RET_ERROR;
+  }
+
+  ParameterDescriptor * parameter_descriptor = &parameter_server->parameter_descriptors.data[index];
+
+  if (parameter_descriptor->type != RCLC_PARAMETER_INT) {
+    return RCL_RET_ERROR;
+  }
+
+  parameter_descriptor->integer_range.data->from_value = from_value;
+  parameter_descriptor->integer_range.data->to_value = to_value;
+  parameter_descriptor->integer_range.data->step = step;
+  parameter_descriptor->integer_range.size = 1;
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rclc_parameter_execute_callback(
+  rclc_parameter_server_t * parameter_server,
+  const Parameter * old_param, const Parameter * new_param)
+{
+  bool ret = true;
+
```

Comment:
Please add NULL pointer checks for all pointer arguments.

Reply:
Added a8431b4ed7f6ee02ab27705d86f60b30eae6b052